pilot/pkg/xds/ads.go

```diff
@@ -568,7 +568,7 @@ func (s *DiscoveryServer) initializeProxy(node *core.Node, con *Connection) erro
 	proxy := con.proxy
 	// this should be done before we look for service instances, but after we load metadata
 	// TODO fix check in kubecontroller treat echo VMs like there isn't a pod
-	if err := s.WorkloadEntryController.RegisterWorkload(proxy, con.Connect); err != nil {
+	if err := s.WorkloadEntryController.RegisterWorkload(proxy, node, con.Connect); err != nil {
```

Comment:
so seems like we are passing `node` here since we haven't yet computed proxy.Locality (L580 and on). Would it work to just re-arrange the order?

Reply:
IIUC no, because L580 gets proxy.Locality from the service instances, which is derived from the autoregistered WorkloadEntry. Hence the comment ```// this should be done before we look for service instances, but after we load metadata```

Reply:
Makes sense. Wondering about this comment:

https://github.com/istio/istio/blob/245872d64a87675a40c542efb8374a3126900a3a/pilot/pkg/xds/ads.go#L580-L581

1. Why do we expect this? Isn't the point that different services may have different locality? 
2. If this expectation is true, why does it break in the case for this PR.

Reply:
Thanks for looking @stevenctl.

Re 1:

```
// ServiceInstance represents an individual instance of a specific version
// of a service. It binds a network endpoint (ip:port), the service
// description (which is oblivious to various versions) and a set of labels
// that describe the service version associated with this instance.
//
// Since a ServiceInstance has a single IstioEndpoint, which has a single port,
// multiple ServiceInstances are required to represent a workload that listens
// on multiple ports.
```
Source: https://github.com/airbnb/istio/blob/master/pilot/pkg/model/service.go#L210-L217

IIUC, a proxy is associated with one or more ServiceInstances, but proxy should only have a single IP, so all the ServiceInstances associated would also have the same IP.

Re 2:

My understanding is that in the case of Autoregistration, ServiceInstance is derived from 1) the auto-registered WorkloadEntry, and 2) the Service/ServiceEntry selecting the WorkloadEntry. So before autoregistration finishes creating the WorkloadEntry, there are no associated ServiceInstances with the proxy, and thus no locality to set.

Reply:
Actually, I think it's possible for a proxy's ServiceInstances to have different IPs, but it seems reasonable that a single running proxy should not be associated with multiple localities, since locality is meant (IMO) to model physical proximity rather than logical topology.

Reply:
This shouldn't matter, since the ServiceInstance won't be there until the WorkloadEntry is created.

Reply:
Yep. So I think the expectation of "all instances [of a proxy] to have ... the same locality" still holds after this PR. Let me know if I misunderstood your questions.